Support v4 mapping to v6

jukkar opened the pull request:
This PR introduces v4-mapping-to-v6 and properly implements IPV6_V6ONLY socket option.
The echo sample application is converted to use this feature.

jukkar committed:
net: Add support for v4-mapping-to-v6 sockets

This allows IPv4 and IPv6 share the same port space.
User can still control the behavior of the v4-mapping-to-v6
by using the IPV6_V6ONLY socket option at runtime.
Currently the IPv4 mapping to IPv6 is turned off by
default, and also the IPV6_V6ONLY is true by default which
means that IPv4 and IPv6 do not share the port space.
Only way to use v4-mapping-to-v6 is to enable the Kconfig
option and turn off the v6only socket option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: Add tests for IPV6_V6ONLY option

The IPV6_V6ONLY option is only checked if
CONFIG_NET_IPV4_MAPPED_TO_IPV6 option is set.
By default the IPV6_V6ONLY option is set.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -78,6 +78,15 @@ bool net_context_is_reuseport_set(struct net_context *context)
+#if defined(CONFIG_NET_IPV4_MAPPING_TO_IPV6)
+	return context->options.ipv6_v6only;
+#else
+	return true;
```
Missing `ARG_UNUSED(context);`

pdgendt commented on subsys/net/ip/Kconfig:
```diff
@@ -143,6 +143,18 @@ source "subsys/net/ip/Kconfig.ipv6"
+	depends on NET_NATIVE_IPV6
+	help
+	  Support v4-mapped-on-v6 address type. This allows IPv4 and IPv6
+	  to share a local port space. When application gets an IPv4
```
```suggestion
	  to share a local port space. When the application gets an IPv4
```

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -1496,6 +1513,32 @@ static int get_context_reuseport(struct net_context *context,
+		return -EINVAL;
+	}
+
+	if (context->options.ipv6_v6only == true) {
+		*((int *)value) = (int) true;
+	} else {
+		*((int *)value) = (int) false;
+	}
```
```suggestion
	*((int *)value) = (int)context->options.ipv6_v6only;
```

jukkar replied:
I decided not to do this in order this function look similar as the other get socket option functions.

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -1496,6 +1513,32 @@ static int get_context_reuseport(struct net_context *context,
+
+	return 0;
+#else
+	return -ENOTSUP;
```
Missing `ARG_UNUSED(context);`, `ARG_UNUSED(value);` and `ARG_UNUSED(len);`

pdgendt reviewed (commented):
Thanks for this! Some minor remarks after a short review.

jukkar committed:
net: tcp: Set address len properly when accepting a socket

The address length of the accepted socket must reflect the
size of the address struct, so it should either be
sizeof(struct sockaddr_in) for IPv4 or sizeof(struct sockaddr_in6) for
IPv6 socket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: utils: Add helpers for manipulating IPv4 mapped address

Create a function that can create IPv4 mapped to IPv6 address.
Create a function that can check if the IPv6 address is a
mapped IPv4 one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: utils: Print IPv4 mapped IPv6 addresses properly

Add support for IPv4 mapped IPv6 addresses when converting
IP address to a string in inet_ntop() function.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: tcp: Fix the accepted socket address

The socket address passed in accept() call should point
to the new connection address and not the old one.
Fortunately this only affects things after the v4-mapping-to-v6
support so older code than this works fine.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: sockets: echo: Convert to work both IPv6 and IPv4

The sample enables CONFIG_NET_IPV4_MAPPING_TO_IPV6 and then
turns off IPV6_V6ONLY option which allows it to support both
IPv6 and IPv4 using the same socket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: conn: Fix the rank value print

We should use the NET_CONN_RANK() macro when printing the
current rank value as that macro masks the rank values properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: misc: Fix memory leak

The sent pkt must be removed after we have cloned it
in the simulated driver otherwise there is a memory leak.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
@kartben The documentation build complains about this comment in the code
```
 *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
 *         look like ::ffff:a.b.c.d
```
The warning is this one
```
WARNING: /runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:1258: explicit link request to 'ffff' could not be resolved
```
Is there a way to mark this in the code so that doxygen does not think it is a link?

kartben commented on include/zephyr/net/net_ip.h:
```diff
@@ -1239,6 +1239,39 @@ static inline void net_ipv6_addr_create_ll_allrouters_mcast(struct in6_addr *add
+
+/**
+ *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
+ *         look like ::ffff:a.b.c.d
```
The issue is with Sphinx, that thinks it's a directive. The following should do the trick (and it's not a bad idea anyway to format the IP address as code, I guess?)

```suggestion
 *         look like `::ffff:a.b.c.d`
```

jukkar replied:
I figured it out, using `\` worked
```
diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
index 20d95d68b8..bc760e7323 100644
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -1255,7 +1255,7 @@ static inline void net_ipv6_addr_create_v4_mapped(const struct in_addr *addr4,
 
 /**
  *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
- *         look like `::ffff:a.b.c.d`
+ *         look like \::ffff:a.b.c.d
  *
  *  @param addr IPv6 address
  *
```

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -1726,8 +1775,21 @@ static int context_sendto(struct net_context *context,
 
 	} else if (IS_ENABLED(CONFIG_NET_IPV4) &&
 		   net_context_get_family(context) == AF_INET) {
```
Is this the right place for the translation? This part of code is covering IPv4 sockets (AF_INET), I had an impression those should still cover IPv4 communications only? Shoudn't the IPv6 code (above) check for IPv4 mapped address?

jukkar replied:
When the socket is accepted and if the incoming data is IPv4, then the accepted net_context family is set to IPv4 even if the listening one is IPv6 and the application socket is IPv6. It looks like a big mess if we try to handle IPv4 data in IPv6 path in other parts of the network stack. With these minimal changes, the echo sample application I modified works fine both in Zephyr and Linux.

rlubos commented on tests/net/socket/misc/src/main.c:
```diff
@@ -610,4 +864,11 @@ ZTEST_USER(socket_misc_test_suite, test_ipv6)
 
+ZTEST_USER(socket_misc_test_suite, test_ipv4_mapped_to_ipv6)
+{
+	test_ipv4_mapped_to_ipv6_disabled();
+	test_ipv4_mapped_to_ipv6_enabled();
+	test_ipv4_mapped_to_ipv6_server();
```
Do those have to execute sequentially, i. e. couldn't they be separate test cases?

jukkar replied:
They could be separate as they do not depend on each other. I can change this if needed.

jukkar committed:
net: tcp: Do not overwrite remote address in new conn

When a new incoming connection is accepted, do not overwrite
the listening remote address. The remote address for the
listening socket is the any address (like :: for IPv6)
and not the accepted socket remote address.
This only affects the "net conn" shell command which prints
the remote address incorrectly. There is no problem accepting
new connections in this case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Print v4-mapping-to-v6 address properly

The remote address of the connection is checked whether
it is v4-mapping-to-v6 address in which case we should
print it such.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/ip/connection.c:
```diff
@@ -380,6 +380,12 @@ int net_conn_register(uint16_t proto, uint8_t family,
 	conn_set_used(conn);
 
+	if (net_context_is_v6only_set(context)) {
+		conn->v6only = true;
```
nit: Maybe just assign the result directly?
```
conn->v6only = net_context_is_v6only_set(context);
```

rlubos commented on subsys/net/ip/connection.c:
```diff
@@ -740,7 +756,24 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
+					continue; /* wrong local address */
+				}
+
+				/* We might have a match for v4-to-v6 mapping, check even more */
```
I'm not sure I understand this comment, `check even more` what?

jukkar replied:
Heh, that was a leftover comment from when testing the connection logic, I will tweak this to be more sane.

rlubos reviewed (commented):
Any clue what's wrong with the new testcase?

I must say this mapping seems to me as a bit tricky topic, but on the other side the echo sample seems to work fine now with both IPv4 and IPv6 and we get some extra test coverage, so fine by me.

jukkar commented:
The net-shell refactoring PR https://github.com/zephyrproject-rtos/zephyr/pull/63164 should be merged before this one so adding DNM

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -2591,6 +2653,28 @@ static int set_context_reuseport(struct net_context *context,
+
+	return 0;
+#else
+	return -ENOTSUP;
```
```suggestion
	ARG_UNUSED(context);
	ARG_UNUSED(value);
	ARG_UNUSED(len);

	return -ENOTSUP;
```

kartben commented:
@jukkar It feels like the README for the sample should be updated?
